vscode/src/ghCopilot.ts

```diff
@@ -0,0 +1,226 @@
+                  "ref": "v1.15.0",
+                  "owner": "microsoft",
+                  "repo": "qsharp",
+                  "path": "library/chemistry"
+              }
+          }
+      }
+  }
+  \`\`\`
+
+  ## Jupyter Notebooks
+
+  Q# has first-class support for Jupyter notebooks. Typically the first cell will contain \`import qsharp\`.
+
+  Jupyter cells can contain Q# code directly by using the \`%%qsharp\` magic command at the beginning of the cell. For example:
+
+  \`\`\`python
+  %%qsharp
+
+  operation GHZSample(n: Int) : Result[] {
+      use qs = Qubit[n];
+
+      H(qs[0]);
+      ApplyToEach(CNOT(qs[0], _), qs[1...]);
+
+      let results = MeasureEachZ(qs);
+      ResetAll(qs);
+      return results;
+  }
+  \`\`\`
+
+  The \`qsharp_widgets\` package provides viewers for circuits and histograms, e.g.
+
+  \`\`\`python
+  from qsharp_widgets import Circuit, Histogram
+  Circuit(qsharp.circuit("GHZSample(3)"))
+  \`\`\`
+
+  Note that the latest Q# and QDK releases don't require or use the old IQ# kernel. It just needs to the \`qsharp\` PyPI package,
+  and maybe \`qsharp_widgets\` for visuals.`;
+
+/**
+ * Checks if the copilot-instructions.md file exists and contains Q# instructions.
+ *
+ * @param workspaceFolder The workspace folder URI.
+ * @returns A promise that resolves to true if the file exists and contains Q# instructions.
+ */
+async function hasQSharpCopilotInstructions(
+  workspaceFolder: vscode.Uri,
+): Promise<boolean> {
+  try {
+    const githubDir = vscode.Uri.joinPath(workspaceFolder, ".github");
+    const instructionsFile = vscode.Uri.joinPath(
+      githubDir,
+      "copilot-instructions.md",
+    );
+
+    // Check if file exists
+    await vscode.workspace.fs.stat(instructionsFile);
+
+    // Check if file contains Q# instructions
+    const existingContent =
+      await vscode.workspace.fs.readFile(instructionsFile);
+    const existingText = new TextDecoder("utf-8").decode(existingContent);
+
+    return existingText.includes(codingInstructionsTitle);
+  } catch {
+    // If any error occurs (file not found, etc.), return false
+    return false;
+  }
+}
+
+/**
+ * Command to update or create the Copilot instructions file for Q#.
+ * Shows a prompt to the user and updates the file if confirmed.
+ */
+async function updateGhCopilotInstructionsCommand() {
+  const workspaceFolders = vscode.workspace.workspaceFolders;
+  if (!workspaceFolders || workspaceFolders.length === 0) {
+    vscode.window.showErrorMessage("No workspace folder is open");
+    return;
+  }
+
+  for (const folder of workspaceFolders) {
+    // Check if the file already exists with Q# instructions
+    const hasInstructions = await hasQSharpCopilotInstructions(folder.uri);
+    if (hasInstructions) {
+      vscode.window.showInformationMessage(
+        "copilot-instructions.md already contains Q# instructions",
+      );
+      return;
+    }
+  }
+
+  const workspaceFolder = workspaceFolders[0].uri;
```

Comment:
Add a TODO here to determine which workspace folder to add the instructions to. Just picking the first open folder if they have multiple isn't correct.

Reply:
Hmm, are you thinking of a definite way we could check, or more like heuristics? We could prefer workspaces that have Q# files (or the one that has the Q# file that's currently open), and fall back to the first one if we can't find a preferred workspace. Something like that? I'll add a todo.

Reply:
the package.json only makes this command available if the resource is a .qs or qsharp.json file, so I assume the user is focused on file in a project/workspace for this to be available? (Likewise if invoked via creating a new project - that's done in the context of a specific folder/workspace)

Reply:
The second param for registerCommand is an optional Uri for the resource it was invoked for (see the `async (folderUri: vscode.Uri | undefined) => {`) code where the `create project` command is registered).

Reply:
VS Code doesn't pass this when invoked from the command palette, currently. Investigating. I can just as well use the "get currently active Q# program" function we use everywhere else for our commands (QIR generation etc)

Reply:
done.